[NORI] Vendingmachine step3

NORIKIM opened the pull request:
자판기 스탭3. 자판기 기본 동작 완료하였습니다.

NORIKIM committed:
고유 특징 프로퍼티 추가

NORIKIM committed:
유효기간을 형식 수정

NORIKIM committed:
Beverage Class 수정 및 추가

상품명과 상품가격을 가지는 변수 생성
유통기한을 확인하는 메소드 validate 수정

NORIKIM committed:
사용자에게 메뉴를 선택하도록 하는 inputView 추가

NORIKIM committed:
main 프로세스 수정

1. 현재 잔액 상태 출력
2. 현재 자판기 내 음료,금액,재고 출력
3. 사용자 선택 메뉴 출력
4. 메뉴 선택하도록 입력 받게 함
5. 선택 지에 따라 값 넘김

NORIKIM committed:
OutputView 생성

1. 현재 투입금액에 대한 문구 출력해주는 메소드(currentStatus) 추가
2.  메뉴 문구 출력해주는 메소드(menu) 추가
3. 현재 보유중인 음료 정보 출력해주는 메소드(beverageList) 추가
4. 사용자의 입력에 오류 발생 시 보유줄 에러 enum errorMessage 추가

NORIKIM committed:
VendingMachine 구조체 생성

1. 음료 인스턴스 생성 - beverage 배열 변수
2. beverage변수를 이용하여 현재 보유중인 음료를 사용자에게 보여주도록 리턴 해주는 메소드(drinks) 추가
3. 현재 잔액에 대한 상태 리턴 해주는 메소드(balance) 추가

NORIKIM committed:
생성된 음료의 재고 상태를 한번에 파악 하도록 딕셔너리 생성(currentBeverage)

1. 딕셔너리의 key : 순서를 정해주기 위한 Int타입 값
2. 딕셔너리의 value : (음료명,가격,개수)형태의 튜플 값

NORIKIM committed:
CurrentBeverage의 key 이용하여 순서대로 현재 음료 상태 출력하도록 수정

NORIKIM committed:
수정된 inventory메소드 반영

NORIKIM committed:
자판기 정의 수정 및 추가

1. 구매 내역을 저장하는 purchase 변수 추가
2. 음료 구매 시 재고,잔액 빼주고 purchase로 음료명 추가해주는 메소드 추가
3. 음료의 유통기한과 따듯한 음료인지 확인 하는 메소드 추가

NORIKIM committed:
출력 메소드 추가

1. 음료 구매 후 어떤 음료를 구매했고 얼마를 차감한다는 안내문구 출력 - printPurchase
2. 에러 발생 시 에러메시지 출력 - printError

NORIKIM committed:
에러의 처리

1. 에러 문구 생성 - enum Error
2. 에러 문구를 리턴 해주는 메소드(Error.message) 생성
3. 초기 입력값 검사 메소드(inCorrect)생성
4. 음료 구매 시 잔액 확인 메소드(notEnoughBalance) 생성

NORIKIM committed:
메인 수정

1. 메인 실행 방법 변경 while -> function
2. 에러 검사 후 출력 부 추가

NORIKIM committed:
선택한 음료의 구매 가능한 재고가 있는지 확인하는 메소드(ourOfStock) 생성

NORIKIM committed:
재고 확인 예외 처리 추가

NORIKIM committed:
에러가 없을 때만 출력하도록 수정

NORIKIM committed:
Unit test VendgingMachine

NORIKIM committed:
Error case Rename inCorrect -> incorrect

NORIKIM committed:
Static 제거

NORIKIM committed:
특정 음료를 구분하는 메소드와 sell메소드에서 불리는 메소드 private으로 변경

NORIKIM committed:
VendingMachine 코드

godrm commented on VendingMachine/VendingMachine/Error.swift:
```diff
@@ -0,0 +1,52 @@
+            guard status[drinks.key]!.2 != 0 else { return .outOfStock }
+        }
+    }
+    return .notError
```
보통 Error 자체를 리턴 타입으로 사용하지는 않습니다. 
최근에 추가된 Result를 사용하기도 합니다만
이 함수가 성공인지 실패인지 구분할 수 있는 상태가 몇 가지인지 생각해서 리턴 타입을 결정해야 합니다.
단지 하나의 상태라면 Bool 만 리턴하고, 그외 나머지 예외적인 경우를 throws 하는 게 좋습니다. 
여기 있는 두 함수는 Error 대신 리턴 타입을 바꿔보세요.

godrm commented on VendingMachine/VendingMachine/VendingMachine.swift:
```diff
@@ -0,0 +1,111 @@
+    
+    // 자판기 금액을 변경해주는 메소드
+    mutating func insert(money: String) {
+        let won = Int(money.dropFirst(2))!
```
money가 문자열인 건 어색하네요. 돈을 가장 잘 표현할 단위는 뭘까요?

godrm commented on VendingMachine/VendingMachine/VendingMachine.swift:
```diff
@@ -0,0 +1,111 @@
+    mutating func inventory() {
+        for drinks in 0 ..< beverage.count {
+            self.currentBeverage.updateValue((beverage[drinks].beverageName,beverage[drinks].beveragePrice,10), forKey: drinks)
+        }
```
이 함수는 return 값이 없어도 되나요?

godrm commented on VendingMachine/VendingMachine/VendingMachine.swift:
```diff
@@ -0,0 +1,111 @@
+        }
+    }
+    // 현재 음료 현황 리턴
+    func currentBeverageStatus() -> [Int:(String,Int,Int)] {
```
리턴 타입이 복잡한데 의미를 모르겠어요. 이런 경우는 저런 값을 포함하는 객체를 추상화해서 새로 선언하는 게 좋습니다.

godrm commented on VendingMachine/VendingMachine/VendingMachine.swift:
```diff
@@ -0,0 +1,111 @@
+    /// 음료수 구매
+    // 구매한 음료 처리
+    mutating func sell(beverage: String) {
+        let selectBeverageNum = Int(beverage.dropFirst(2))! - 1
```
여기 음료값에 대한 것도 문자열이 아니라 음료 메뉴에 대한 값을 다른 타입으로 개선해보세요. 
입력한 값을 메뉴 중에 어떤 것과 일치시키는 메뉴 객체가 있다면 어떨까요?

godrm commented on VendingMachine/VendingMachine/VendingMachine.swift:
```diff
@@ -0,0 +1,111 @@
+                subtract(from: inner.key)
+                purchaseList(from: inner.key)
+                deduct(money: inner.key)
+                OutputView().printPurchase(productName: currentBeverage[inner.key]!.0, price: currentBeverage[inner.key]!.1)
```
위에 했던 피드백과 이어질 텐데 currentBeverage를 추상화한 객체가 있다면 여기 로직이 훨씬 단순해질 것 같습니다. 
VendingMachine 에서 직접 OutputView를 호출하면 안됩니다. 
값을 가져간 곳에서 출력하도록 개선하세요.

godrm commented on VendingMachine/VendingMachine/VendingMachine.swift:
```diff
@@ -0,0 +1,111 @@
+    }
+
+    // 따뜻한 음료만 리턴하는 메소드
+    private func hotBeverage() -> [String] {
```
재고나 음료를 리턴할 때는 `[String]` 대신 `[Beverage]` 형식으로 음료수 객체를 직접 다루는 게 좋습니다.

godrm commented on VendingMachine/VendingMachine/main.swift:
```diff
@@ -8,10 +8,26 @@
+        case "1": vendingMachine.insert(money: input)
+        case "2":
+            guard notEnoughBalance(of: vendingMachine, input) == .notError else { return OutputView().printError(notEnoughBalance(of: vendingMachine, input)) }
+            guard outOfStock(machine: vendingMachine, input) == .notError else { return OutputView().printError(outOfStock(machine: vendingMachine, input))  }
```
아마 여기 함수들 리턴 타입이 달라지면 여기 코드도 좀 달라질 겁니다. try - catch를 활용해보세요.

godrm commented on VendingMachine/VendingMachine/main.swift:
```diff
@@ -8,10 +8,26 @@
+    }
+}
+while true {
+    main()
```
메인을 반복하지 말고 main 내부에서 반복문을 사용하세요.

NORIKIM committed:
insert메소드 파라미터 타입 변경 String -> Int

NORIKIM committed:
UnitTest

NORIKIM committed:
예외 처리 파일 수정

1. notError삭제
2. incorrect메소드 - 최초 입력값을 확인하고 통과 시 메뉴와 값(userChoice)으로 구분하여 반환
3. checkAvailability메소드 - 잔액과 재고확인 메소드를 통합

NORIKIM committed:
Inventory객체 생성

1. 메뉴 enum 추가
2. 음료의 이름,가격,재고를 가짐

NORIKIM committed:
수정

1. 복잡한 currentBeverage를 Inventory구조체를 가지도록 함
2. 음료 구매 즉시 내역 출력 해주는 구문 삭제
3. 특정 음료 구분하는 notValiDate, hotBeverage 메소드 리턴 타입 변경 [String] -> [Beverage]

NORIKIM committed:
수정

1. 단순 출력 메소드 static으로 변경
2. 불필요한 구문 삭제

NORIKIM committed:
Do-catch를 이용하여 수정

NORIKIM committed:
VendingMachine code

NORIKIM commented:
피드백 반영 수정하였습니다.

godrm commented on VendingMachine/VendingMachine/Inventory.swift:
```diff
@@ -0,0 +1,27 @@
+    enum Menu: String {
+        case insertMoney = "1"
+        case buyBeverage = "2"
+    }
```
Menu는 Inventory와 관련 있기 보다는 자판기 자체와 관련 있는게 아닐까요?
돈을 넣는 메뉴가 Inventory 내부에 있는게 어색합니다.

godrm commented on VendingMachine/VendingMachine/Inventory.swift:
```diff
@@ -0,0 +1,27 @@
+    
+    var beverageName: String
+    var beveragaPrice: Int
+    var beverageCount: Int
```
객체 인스턴스를 만들어서 Inventory를 구성하는 게 좋습니다.
얼마짜리 딸기 우유가 몇 개 있다는 것을 정보로 구성하는 게 아니라,
딸기우유 객체 인스턴스를 배열에 넣어서 몇 개가 있는지 확인하는거죠. 
물론 생성할 때는 몇 개를 만들지 받아서 한꺼번에 만들수도 있겠죠.

godrm commented on VendingMachine/VendingMachine/VendingMachine.swift:
```diff
@@ -18,7 +18,7 @@ struct VendingMachine {
     }
     private var currentBalance: Int
-    private var currentBeverage = [Int:(String,Int,Int)]()
+    private var currentBeverage = [Inventory]()
```
Inventory는 음료 객체 인스턴스를 관리하는 객체가 되어야 합니다.

godrm commented on VendingMachine/VendingMachine/VendingMachine.swift:
```diff
@@ -43,69 +42,58 @@ struct VendingMachine {
+        let coffee = beverage.filter{
+            let drink = $0 as? Coffee
+                if drink?.isHot() == true { return true }
+            return false
```
if 구문 내부에 let 바인딩과 isHot() 호출을 모두 판단해도 되지 않았을까요?
혹은 guard let 과 isHot() 호출도 표현할 수 있을 것 같습니다.

이런 바인딩과 비교는 자주 사용하게 될텐데 읽기쉽고, 안전하게 처리하는 것을 늘 먼저 고민하세요. 
물론 지금도 잘 동작합니다만, 들여쓰기도 그렇고 어색한 부분이 있습니다.
`return true` 같은 표현을 직접 쓰기 보다 `return drink?.isHot()` 으로 해도 됩니다.

godrm commented on VendingMachine/VendingMachine/main.swift:
```diff
@@ -11,23 +11,48 @@ import Foundation
+    static func input() throws -> userChoice {
+        OutputView.currentStatus(vendingMachine.balance()) // 현재 투입한 금액이 0원입니다. 다음과 같은 음료가 있습니다.
+        OutputView().beverageList(vendingMachine)
+        OutputView.menu() // 1. 금액추가 2. 음료구매
```
`OutputView().beverageList()` 만 왜 OutputView 인스턴스를 생성하고 있나요? 
일관성 있게 사용할 수 있도록 하는게 좋습니다. 이렇게 쓰면 헷갈릴 수 있습니다.

godrm reviewed (commented):
많이 개선되었네요. 우선 이번 단계에서 마무리하고 피드백 내용을 확인해서 다음 단계를 할 때 같이 개선해도 좋습니다.
